### PR TITLE
add parallel binary edge list shuffler script

### PR DESCRIPTION
**Pull request by meyerzinn**
Test this script with an edge list file:

```bash
$ cat graph.el | tr '\t ' '\n' | perl -pe '$_=pack"Q",$_' > graph.bel
$ ./shufbel -rseed 1 graph.bel graph_s1.bel
$ md5sum graph_s1.bel
<some hash gets printed>
$ rm graph_s1.bel
$ ./shufbel -rseed 1 graph.bel graph_s1.bel
$ md5sum graph_s1.bel
<the same hash gets printed>
```

I also checked that the triangle count did not change on rmat14 vs lonestar triangle counting, which isn't conclusive proof that the edge list is the same but is pretty strong evidence.

### CODE REVIEW

File: scripts/shufbel/shufbel.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+	fout.Truncate(int64(sizeBytes))
+
+	data, err := syscall.Mmap(int(fin.Fd()), 0, int(sizeBytes), syscall.PROT_READ, syscall.MAP_PRIVATE|syscall.MAP_NORESERVE)
+	if err != nil {
+		fmt.Printf("Error mmaping file %s: %v\n", input_file, err)
+		return
+	}
+	if uintptr(unsafe.Pointer(&data[0]))%uintptr(pageSize) != 0 {
+		fmt.Println("mmap returned a non-page-aligned address: %p", uintptr(unsafe.Pointer(&data[0])))
```
Comment (AdityaAtulTewari) [inline scripts/shufbel/shufbel.go:77]: why aren't you dispatching goroutines to do striped reads of the file

**Review by AdityaAtulTewari — APPROVED** [review]

File: scripts/shufbel/shufbel.go (hunk 1 above)
Comment (meyerzinn) [reply scripts/shufbel/shufbel.go:77]: I am, scroll down

File: scripts/shufbel/shufbel.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+	num_edges := sizeBytes / 16
+
+	rand.Seed(*rseed)
+	// keys should be an array of 4 random uint64s
+	keys := []shuffle.FeistelWord{shuffle.FeistelWord(rand.Uint64()), shuffle.FeistelWord(rand.Uint64()), shuffle.FeistelWord(rand.Uint64()), shuffle.FeistelWord(rand.Uint64())}
+
+	var wg sync.WaitGroup
+	// spawn a goroutine for every page, since each page can fault
+	// todo (meyer): should we limit the number of concurrent goroutines?
```
Comment (AdityaAtulTewari) [inline scripts/shufbel/shufbel.go:102]: Not todo, yes. make it a input parameter, also make threads controllable by a input parameter.

File: scripts/shufbel/shufbel.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	for i := uint64(0); i < num_edges; i += uint64(pageSize / 16) {
```
Comment (AdityaAtulTewari) [inline scripts/shufbel/shufbel.go:103]: This needs to also be a program input.

**Review by AdityaAtulTewari — CHANGES_REQUESTED** [review]

**Comment by AdityaAtulTewari** [comment]
please revert the approval was a mistake

File: scripts/shufbel/shufbel.go (hunk 2 above)
Comment (meyerzinn) [reply scripts/shufbel/shufbel.go:102]: threads is already controllable.

File: scripts/shufbel/shufbel.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math/rand"
```
[the remark this answers is not in the document]
Comment (meyerzinn) [reply scripts/shufbel/shufbel.go:6]: It is called once to generate 4 ints to be the feistel keys.

File: scripts/shufbel/shufbel.go (hunk 1 above)
Comment (AdityaAtulTewari) [reply scripts/shufbel/shufbel.go:77]: No I specifically mean like a read command, this is supposed to work on graphs that are TB in size. How will this even be mmaped?

File: scripts/shufbel/shufbel.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply scripts/shufbel/shufbel.go:6]: oh that's fine, Idk why I thought that was in the loop.

**Pull request merged**